cli/src/main.rs

```diff
@@ -12,14 +12,70 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::sync::OnceLock;
+
 #[cfg(feature = "mimalloc")]
 use jj_cbits::mimalloc::MiMalloc;
 use jj_cli::cli_util::CliRunner;
+use jj_cli::command_error::CommandError;
+
+/// Lazy global static. Used only to defer printing mimalloc stats until the
+/// program exits, if set to `true`.
+static PRINT_HEAP_STATS: OnceLock<bool> = OnceLock::new();
+
+#[derive(clap::Args, Clone, Debug)]
+pub struct ShowAllocStats {
+    /// Show memory allocation statistics from the internal heap allocator
+    /// on `stderr`, when the program exits.
+    #[arg(long, global = true)]
+    show_heap_stats: bool,
+}
+
+/// Enable heap statistics for the user interface; should be used with
+/// [`CliRunner::add_global_args`]. Does nothing if the memory allocator is
+/// unused, i.e. `#[global_allocator]` is not set to mimalloc in your program.
+pub fn heap_stats_enable(
+    _ui: &mut jj_cli::ui::Ui,
+    opts: ShowAllocStats,
+) -> Result<(), CommandError> {
+    if opts.show_heap_stats {
+        PRINT_HEAP_STATS.set(true).unwrap();
+    }
+    Ok(())
+}
 
 #[cfg(feature = "mimalloc")]
 #[global_allocator]
 static ALLOC: MiMalloc = MiMalloc;
 
 fn main() -> std::process::ExitCode {
-    CliRunner::init().version(env!("JJ_VERSION")).run()
+    let result = CliRunner::init()
+        // NOTE (aseipp): always attach heap_stats_enable here, even if compiled
+        // without mimalloc; we don't want the test suite or other users to have
+        // to worry about if the command exists or not based on the build
+        // configuration
+        .add_global_args(heap_stats_enable)
```

Comment:
This shouldn't need global state, especially since we read and consume the argument in the same function. `CliRunner` should probably be refactored to take a lifetime, but even if you don't want to take that on, it would be cleaner to use an `Arc<AtomicBool>` here.